static-site/src/components/ListResources/useDataFetch.ts

```diff
@@ -56,55 +62,56 @@ export function useDataFetch(sourceId, versioned, defaultGroupLinks, groupDispla
  * representing group names (pathogen names) and values which are arrays of
  * resource objects. 
  */
-function partitionByPathogen(pathVersions, pathPrefix, versioned) {
+function partitionByPathogen(pathVersions: PathVersions, pathPrefix: string, versioned: boolean) {
+  return Object.keys(pathVersions).reduce((store: Partitions, name: string) => {
+    // name will always be a valid key based on reduce() above
+    const sortedDates = [...pathVersions[name]!].sort();
 
-  return Object.entries(pathVersions).reduce(reduceFn, {});
-  
-  function reduceFn(store, datum) {
-    const name = datum[0];
-    const nameParts = name.split('/');
-    const sortedDates = [...datum[1]].sort();
+    // do nothing if resource has no dates
+    if (sortedDates.length < 1) return store
 
-    let groupName = nameParts[0];
+    const nameParts = name.split('/');
+    // split() will always return at least 1 string
+    let groupName = nameParts[0]!;
 
-    if (!store[groupName]) store[groupName] = []
-    const resourceDetails = {
+    const resourceDetails: Resource = {
       name,
       groupName, /* decoupled from nameParts */
       nameParts,
       sortingName: _sortableName(nameParts),
       url: `/${pathPrefix}${name}`,
-      lastUpdated: sortedDates.at(-1),
+      lastUpdated: sortedDates.at(-1)!,
       versioned
     };
     if (versioned) {
-      resourceDetails.firstUpdated = sortedDates[0];
-      resourceDetails.lastUpdated = sortedDates.at(-1);
+      resourceDetails.firstUpdated = sortedDates[0]!;
+      resourceDetails.lastUpdated = sortedDates.at(-1)!;
       resourceDetails.dates = sortedDates;
       resourceDetails.nVersions = sortedDates.length;
       resourceDetails.updateCadence = updateCadence(sortedDates.map((date)=> new Date(date)));
     }
 
-    store[groupName].push(resourceDetails)
+    if (!store[groupName]) store[groupName] = [];
+    store[groupName]!.push(resourceDetails)
 
     return store;
-  }
+  }, {});
 }
 
 
 /**
  * Turn the provided partitions (an object mapping groupName to an array of resources)
  * into an array of groups.
  */
-function groupsFrom(partitions, pathPrefix, defaultGroupLinks, groupDisplayNames) {
-  return Object.entries(partitions).map(makeGroup)
-
-  function makeGroup([groupName, resources]) {
-    const groupInfo = {
+function groupsFrom(partitions: Partitions, pathPrefix: string, defaultGroupLinks: boolean, groupDisplayNames: GroupDisplayNames) {
+  return Object.keys(partitions).map(groupName => {
+    // groupName will always be a valid key based on map() above
+    const resources = partitions[groupName]!;
```

Comment:
[not a request for changes] I guess the more ergonomic way is something like the following?
```js
Object.entries(partitions).map(([groupName, resources]) => {
```
Is this assertion needed because the compiler isn't smart enough to know that group name is taken from the valid keys of partitions, or because mutability means that a key may have been removed from `partitions` before the map function gets that key?

Reply:
It's simply because operator [] returns either `T` or `undefined`. It does not care about any other conditions.

Reply:
@jameshadfield I got confused and didn't realize that [groupName, resources] is properly typed as long as the function is defined inline within `Object.entries().map()`. Changed 2aeb155851e225c5039fa172d96f681f6c665fba to 34a314070e2d1326555a5df78323f8e8086c9dc2 which removes the need for the bang (reflected in [force-push](https://github.com/nextstrain/nextstrain.org/compare/aae874c06cd6d17616710c501ad73104a7c5e051..7c750e278da27459f54a91ee373580e33089777d)).